Bolts/Common/BFTask.m

```diff
@@ -430,14 +434,7 @@ - (void)waitUntilFinished {
         [self warnOperationOnMainThread];
     }
 
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return;
-        }
-        [self.condition lock];
-    }
-    [self.condition wait];
-    [self.condition unlock];
+    dispatch_semaphore_wait(_completedSemaphore, DISPATCH_TIME_FOREVER);
```

Comment:
There is an issue with this one.
`waitUntilFinished` should be a no-op if the task is already completed, dependless of how many times it is called.
The previous flow had a check for that around `if (self.completed) return`, this one doesn't.

I think simply adding lines 433-436 back would solve the problem.